### PR TITLE
fix: enforce undelegation over delegation guard

### PR DESCRIPTION
**Pull request by charmful0x**
### issue
pot undelegation was missing a delegation bound check (`Amount <= ExistingDelegation`) inside `undelegate/6` which would have let a delegator to undelegate more than his ExistingDelegation (up from the upsteam undelegation msg flow). 

and true there is a `undelegate_more_than_delegated_test` test, but it was failing on a different path, `liquidate` instead of over undelegation guard check (which was skipped as it didnt exist). therefore that test was an edge case test instead of classic undelegation test (undelegating more than what is delegated) - failing on liquidation throw

the original flow was: `RecipientDeposit -> liquidate/5 -> drip -> state transitions (S)`

### fix

the fix introduced in this PR does a pre-validation in `undelegate/6` before going in the original flow on success: `Validation (non-self over-undelegation?) -> RecipientDeposit -> liquidate -> drip -> Ss`

validation returns `{error, <<"Undelegation amount exceeds existing delegation.">>}` on failure

additionally `undelegate/3` has been improved to return `{ok, NewState}` on sucess and `{error, Reason}` on failure, as same as request in #665 review

### tests

i added `undelegate_exceeds_delegated_rejected_with_recipient_buffer_test` - this case was previously unprotected and is now rejected

```
  [done in 2.685 s]
=======================================================
  All 64 tests passed.
```

i will do the response-shapre alignment cleanup in the next PR

### CODE REVIEW

**Comment by charmful0x** [comment]
added a commit that makes sense under the scope of this PR/branch -  align liquidation error paths in upstream deposit/withdraw handling

handled edge case: via upstream actions a user can send over-withdraw reqs without any prior delegation state, and when a withdraw overdraws the user deposit, `withdraw/5` attempts to unwind outgoing delegations to fulfill the over-withdraw. if delegations are empty/0, the old path could crash on `lists:max([])` <- edge case

the fix returns a tuple error `({error, <<"Insufficient delegated balance to liquidate.">>})` instead of throwing and propagates it upstreamly via `withdraw/5` and `undelegate/6`, and aligning with upstream token error handling

File: src/dev_pot.erl [hunk 1]
```diff
@@ -695,41 +672,118 @@ undelegate(FromAddr, ToAddr, ResourceID, Amount, S, Opts) when Amount > 0 ->
...
-                    ToAddr,
-                    ResourceID,
-                    get_deposit(ToAddr, ResourceID, S4, Opts),
-                    S5,
-                    Opts
-                ),
-            send_delegation_notice(FromAddr, ToAddr, ResourceID, -Amount, S6, Opts)
+                )
+        end,
+    Validation =
```
Comment (samcamwilliams) [inline src/dev_pot.erl:699]: What do you think about trying to use the `maybe ... end` pattern for error propagation in this/these modules? Experience so far has been that if we use ok|error tuples it works nicely because the errors always fall through and propagate upwards, but in cases where we have bare returns or errors (`NewStateMsg | {error, Err}` or `#{ <<"status">> => ErrCode, ...}`) it is harder to make use of the pattern matching. Looks like we have some of those in this module so we would have to unify on the `{ok|error, ...}` style if we wanted to `maybe`ify it.

File: src/dev_pot_test_vectors.erl [hunk 2]
```diff
@@ -948,7 +951,23 @@ undelegate_more_than_delegated_test() ->
     S0 = pot_state_multi(ResourceOxygen, [{Alice, 10}, {Bob, 0}]),
     S1 = dev_pot:delegate(Alice, Bob, ResourceOxygen, 5, S0, Opts),
     % Delegation record shows 5, try to undelegate 10
-    ?assertError(_, dev_pot:undelegate(Alice, Bob, ResourceOxygen, 10, S1, Opts)).
+    ?assertMatch(
```
Comment (samcamwilliams) [inline src/dev_pot_test_vectors.erl:966]: Nice catch 🫡

File: src/dev_pot.erl (hunk 1 above)
Comment (charmful0x) [reply src/dev_pot.erl:699]: thanks for the review! i addressed this in the latest commits :saluting_face: 

to keep this PR in scope, i standardized return shape for `delegate/6`, `undelegate/6`, and their `/3` wrappers, and refactored to use `maybe`ied for error propagation

also i added type aliases and specs for the main exported functions (`delegate`, `undelegate`, `deposit`, `withdraw`), and updated tests to match the new `delegate/undelegate` return shapes

wrt `deposit`/`withdraw` - they arent fully aligned yet, since normalizing those return shapes would require a broader refactor touching tests and the upstream `dev_token` mod. 

Q1: is it fine to do that in a follow up cleanup PR? for now, despite mixed return shapes there, i added specs to make that cleanup easier

Q2 on spec style: i used local aliases for domain specific types (`addr`, `resource_id`, `pot_state`, `pot_reason`) and inline types for generic args (`Opts`, `Assignment`, etc). would you prefer full inline style (like `dev_green_zone`), or is this approach acceptable?

Q3: are there plans to have codebase-wide shared type aliases/specs for common generic shapes? (e.g. request/assignment/opts/result), or should i keep those local+inline per module?
